### PR TITLE
Update tools for hydrated test

### PR DESCRIPTION
**Pull request by 0xba1a**
When focused on testing the RPMs, we don't need to follow normal build
procedure. So updating the toolchain for hydrated test in below ways,

 * Remove unresolved dependency nodes. Let the build proceed and package
 fail due to dependency failure. We'll get the info from the log
 * Don't rebuild a package if its dependency is built. As the build is
 only for testing it, there won't be changes which will cause the
 dependent package to fail
 * Don't stop the build after a package build failure. For that if a
 package fails to build, record its failure in logs and proceed by
 using its pre-built RPM

 For this purpose, adding a new macro - RPM_HYDRATED_TEST

###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

When focused on testing the RPMs, we don't need to follow normal build
    procedure. So updating the toolchain for hydrated test in below ways,

     * Remove unresolved dependency nodes. Let the build proceed and package
     fail due to dependency failure. We'll get the info from the log
     * Don't rebuild a package if its dependency is built. As the build is
     only for testing it, there won't be changes which will cause the
     dependent package to fail
     * Don't stop the build after a package build failure. For that if a
     package fails to build, recored its failure in logs and proceed by
     using its pre-built RPM

     For this purpose, adding a new macro - RPM_HYDRATED_TEST

###### Does this affect the toolchain?  

**NO**

###### Test Methodology

- Local build

### CODE REVIEW

**Review by PawelWMS — CHANGES_REQUESTED** [review]
@0xba1a, a few remarks:
- I like the idea of being able to decide if we want to rebuild packages if their dependencies have been built. We used to have this in `1.0` but lost it after the big tooling rework. Please use the `REBUILD_DEP_CHAINS` argument we already have in the `Makefile` and in the documentation, since it's already there and some people will be familiar with it.
- I'm not convinced to ignoring missing dependencies and trying to build anyway. This way we have to wait longer for a failing build and lose that additional information we have now, that the build didn't really fail due to spec misconfiguration but because it's missing a dependency. Instead, I think we might benefit more from extending the build summary where for each blocked SRPM we also print a list of dependencies it is missing. Example

    Now:
    ```
    Blocked SRPMs: 
    --> ocaml-mmap-1.1.0-18.cm2.src.rpm
    ```

    Suggestion:
    ```
    Blocked SRPMs: 
    --> ocaml-mmap-1.1.0-18.cm2.src.rpm. Blocked on: [dep_A, dep_b, ...]
    ```
- We already continue the build even if some of the packages fail through the default `STOP_ON_PKG_FAIL=n` argument. I believe that if we bring back the functionality of `REBUILD_DEP_CHAINS`, we should be set. Without it, some packages don't attempt to build because of those failing dependencies rebuilds, which won't be happening for `REBUILD_DEP_CHAINS=y`.

**Comment by thcrain-msft** [comment]
> * I'm not convinced to ignoring missing dependencies and trying to build anyway. This way we have to wait longer for a failing build and lose that additional information we have now, that the build didn't really fail due to spec misconfiguration but because it's missing a dependency.

I'm with Pawel on this one. I'm not sure I'd trust a positive result for packages building in this fashion, either. Some packages rely on implicit dependency detection in configuration scripts for optional dependencies, which could lead to applicable tests being skipped.

**Comment by 0xba1a** [comment]
@PawelWMS @thcrain-msft,

the macro **REBUILD_DEP_CHAINS** served a different purpose. For example if we are rebuild *openssl* setting **REBUILD_DEP_CHAINS=y** will cause all the packages depending on *openssl* to be rebuilt. And this macro is not used anymore. Seems its just a zombie in the Makefile.

the macro **STOP_ON_PKG_FAIL=y** will stop the entire build on the first package build failure itself. Setting it to **"n"** will just proceed with the packages which are not depending on the failed package.

Neither of them serve the purpose I required. Here my target is to run ptest of all the packages that are having check section. The TestRPM pipeline is having the feature to run ptest for only one package. In that case all the dependencies of that package will hydrated from the BuildRPM pipeline. No package other than the one specified will be built. The same can be done for all the packages in the branch. But running this pipeline 500 times for 500 different packages is not practical.

With the changes in this PR, I'm trying to mimic running TestRPM pipeline for each package individually.
 * I removed the unresolved dependency. Otherwise the entire build process will stop at pkgGraph stage itself. Just because one package not having its dependency, all packages' ptest will not run. Like @thcrain-msft mentions it may skip some valid tests, but right now this is the best available option. I believe this will be resolved when we start addressing each package individually.
 * "**STOP_ON_PKG_FAIL=n**" will not help here. Because if a package fails to build, it shouldn't be marked as failed. Instead its RPM in the cache should be used for building all of its dependent packages. **STOP_ON_PKG_FAIL** doesn't provide this support.

**Comment by PawelWMS** [comment]
@0xba1a: sorry for the long wait! First, let me shortly clarify that by asking to use `REBUILD_DEP_CHAINS` I meant reviving it in 2.0, since it seemed like you need its functionality anyway. After the detailed explanation of what you're looking for, however, I believe that might not be necessary (but still very welcome :)).

To your point: if I understand your needs, you'd like to use pre-built RPMs from the "regular" package builds to jump straight to building only the packages with a `%check` section without having to wait for the dependencies (since all of them are already present). If that's correct: love it, sounds like it's going to save everyone a ton of time. I think we should approach doing that a bit differently, though.

With the current implementation we're shredding the dependency graph to pieces, making it no longer reflect the dependencies between packages and rendering it useless for any post-failure analysis. Having to go through build logs to figure out what went wrong is a regression from what the dependency graph offers, and I'd prefer to avoid it. Instead of that, I'd suggest using a special kind of node we've introduced recently to break circular dependencies between pre-built RPMs: `TypePreBuilt`. What this would require:

1. Iterate through each SRPM's build node (type: `TypeBuild`, status: `StateBuild`).
2. For each SRPM build node, look for its pre-built dependencies (function: `pkggraph.IsSRPMPrebuilt`).
3. For each pre-built dependency: clone it, update the clone (type: `TypePreBuilt`, status: `StateUpToDate`).
4. Make all nodes for that SRPM from the 1<sup>st</sup> point depend on this new clone instead of the old dependency.

This way:

- All packages you want to build will start immediately, since their dependencies should be already present.
- At the same time we get a more universal solution, which will also handle cases where only some of the dependencies are present, which will be very useful for people performing local builds.
- The graph still correctly reflects the dependencies between packages.
- Any build errors will still be accurate and help with debugging any build issues.

Look inside the `fixPrebuiltSRPMsCycle` function to see how the steps above are implemented. It's possible you'll be able to re-use a part of that function if we're lucky.

And we can use the `PACKAGE_BUILD_LIST` and `PACKAGE_REBUILD_LIST` arguments to select a subset of packages to build (in your case, the ones with the `%check` section`).

File: toolkit/tools/scheduler/scheduler.go [hunk 1]
```diff
@@ -71,8 +71,9 @@ var (
 
 	ignoredPackages = app.Flag("ignored-packages", "Space separated list of specs ignoring rebuilds if their dependencies have been updated. Will still build if all of the spec's RPMs have not been built.").String()
 
-	pkgsToBuild   = app.Flag("packages", "Space separated list of top-level packages that should be built. Omit this argument to build all packages.").String()
-	pkgsToRebuild = app.Flag("rebuild-packages", "Space separated list of base package names packages that should be rebuilt.").String()
+	pkgsToBuild     = app.Flag("packages", "Space separated list of top-level packages that should be built. Omit this argument to build all packages.").String()
+	pkgsToRebuild   = app.Flag("rebuild-packages", "Space separated list of base package names packages that should be rebuilt.").String()
+	rpmHydratedTest = app.Flag("rpm-hydrated-test", "Is this just a TestRPM build with all RPMs hydrated?").String()
```
Comment (dmcilvaney) [inline toolkit/tools/scheduler/scheduler.go:76]: This should just be a `.Bool()` flag I think, we don't encode anything beyond y/n here. See `$(RUN_CHECK)` and `$(STOP_ON_PKG_FAIL)` for examples of how to pass those in.

File: toolkit/tools/graphpkgfetcher/graphpkgfetcher.go [hunk 2]
```diff
@@ -38,8 +38,9 @@ var (
 	tlsClientCert = app.Flag("tls-cert", "TLS client certificate to use when downloading files.").String()
 	tlsClientKey  = app.Flag("tls-key", "TLS client key to use when downloading files.").String()
 
-	inputSummaryFile  = app.Flag("input-summary-file", "Path to a file with the summary of packages cloned to be restored").String()
-	outputSummaryFile = app.Flag("output-summary-file", "Path to save the summary of packages cloned").String()
+	inputSummaryFile    = app.Flag("input-summary-file", "Path to a file with the summary of packages cloned to be restored").String()
+	outputSummaryFile   = app.Flag("output-summary-file", "Path to save the summary of packages cloned").String()
+	removeUnresolvedDep = app.Flag("remove-unresolved-dep", "Remove a depedency if its not resolvable").String()
```
Comment (dmcilvaney) [inline toolkit/tools/graphpkgfetcher/graphpkgfetcher.go:43]: Also should be a .Bool() flag.

**Comment by dmcilvaney** [comment]
Agree with Pawel re: editing the graph nodes to PreBuilt.

I'm not a huge fan of adding special more special case handling to the scheduler. We already have so many branching paths in there.

We are meant to have all the requirements satisfied since we started hydrated. If we make that a core assumption, we should be able to take a transformed graph like Pawel suggested. For every single build node, replace all its dependencies with prebuilt nodes. If anything isn't pre-built that means we didn't hydrate correctly, I think?

I'm actually curious if we can use the existing caching behaviour to do some of this (I haven't looked at it in detail, fyi). My understanding is that the scheduler won't rebuild existing packages unless we trip the dependent package check. If we add an option to suppress the rebuild of downstream packages, and pass a list of packages with check section to `PACKAGE_REBUILD_LIST` we might get the desired behaviour.

**Comment by dmcilvaney** [comment]
Another option: We add an optional graph processing step.

Take our normal graph:
- Add a 'check' goal node
 - for every package with a %check section:
   - clone its nodes, add its dependencies as prebuilt versions
 - create a subgraph against the check goal
 - pass to the scheduler as normal

In theory every package with a check section should be immediately ready to build, and we should have no dependent package rebuilds because no packages actually depend on each other.

**Comment by dmcilvaney** [comment]
Ok, here are my suggestion: Add a step which converts nodes to the prebuilt nodes (if they are not part of the PACKAGE_REBUILD_LIST). This means any package which is present will be used, even if normally we would like to rebuild it. We can also have a flag to stop dependent packages rebuilding if needed.

If we do that we should be able to leave the scheduler mostly the same.

Personally, I might add an optional extra step in the makefile to do this step, but it could be included in the existing tools if desired.

**Comment by 0xba1a** [comment]
Correct PR - https://github.com/microsoft/CBL-Mariner/pull/1958

**Pull request closed without merging**